src/box/lua/config/instance_config.lua

```diff
@@ -1513,6 +1513,139 @@ return schema.new('instance_config', schema.record({
             value = schema.scalar({type = 'string'}),
         }),
     }),
+    sharding = schema.record({
+        -- Instance vshard options.
+        roles = schema.set({
+            'router',
+            'storage',
+        }),
+        zone = schema.scalar({type = 'integer'}),
+        -- Cluster vshard options.
+        lock = schema.scalar({
+            type = 'boolean',
+            validate = function(_data, w)
+                local scope = w.schema.scope
+                if scope ~= nil and scope ~= 'global' and scope ~= 'replicaset'
+                   and scope ~= 'instance' then
+                    w.error('sharding.lock should be a defined in '..
+                            'replicaset, group or global scope')
+                end
+            end,
```

Comment:
I suspect that this scope validation (and all the similar functions below) doesn't work. There are several reasons.

1. `scope` annotation is present only in the top level record. Please, consider using of `w.computed.annotations.scope`.
2. `data` is not checked against `nil`/`box.NULL`.
3. The `scope` variable check itself strange. The error message suggests that we should allow `global`, `group` and `replicaset`, but actually we allow `global`, `replicaset` and `instance` (and not `group`).
4. I would also note that I don't understand what the 'cluster vshard options' means. I would understand it if the option would be only allowed in the `global` scope.

There is a tricky point in the scope validation. It seems, we should allow `scope == nil` always, because otherwise a cluster config instantiation into instance config would produce invalid instance config. However, it doesn't mean that we can't forbid something in `scope == 'instance'`.